File: kotlin/src/commonMain/kotlin/lib/humanoid/paromat/interpreter/Interpreter.kt
package lib.humanoid.paromat.interpreter

import lib.humanoid.paromat.node.*

class Interpreter: Evaluator<Double> {

    override fun visit(node: Node): Double = when (node) {
        is NumberNode -> visitNumberNode(node)
        is AddNode -> visitAddNode(node)
        is SubtractNode -> visitSubtractNode(node)
        is MultiplyNode -> visitMultiplyNode(node)
        is DivideNode -> visitDivideNode(node)
        is PlusNode -> visitPlusNode(node)
        is MinusNode -> visitMinusNode(node)
    }

    override fun visitNumberNode(node: NumberNode) = node.number

    override fun visitAddNode(node: AddNode) = visit(node.left) + visit(node.right)

    override fun visitSubtractNode(node: SubtractNode) = visit(node.left) - visit(node.right)

    override fun visitMultiplyNode(node: MultiplyNode) = visit(node.left) * visit(node.right)

    override fun visitDivideNode(node: DivideNode) = visit(node.left) / visit(node.right)

    override fun visitPlusNode(node: PlusNode) = visit(node.node)

    override fun visitMinusNode(node: MinusNode) = - visit(node.node)
}

File: python/paromat/interpreter.py
from nodes import *
from values import Number

class Interpreter:

    def visit(self, node: Node) -> Number:
        if isinstance(node, NumberNode):
            return self.visitNumberNode(node)
        elif isinstance(node, AddNode):
            return self.visitAddNode(node)
        elif isinstance(node, SubtractNode):
            return self.visitSubtractNode(node)
        elif isinstance(node, MultiplyNode):
            return self.visitMultiplyNode(node)
        elif isinstance(node, DivideNode):
            return self.visitDivideNode(node)
        elif isinstance(node, MinusNode):
            return self.visitMinusNode(node)
        elif isinstance(node, PlusNode):
            return self.visitPlusNode(node)
        else:
            return None

    def visitNumberNode(self, node: NumberNode) -> Number:
        return Number(node.value)
    
    def visitAddNode(self, node: AddNode) -> Number:
        return Number(self.visit(node.left).value + self.visit(node.right).value)

    def visitSubtractNode(self, node: SubtractNode) -> Number:
        return Number(self.visit(node.left).value - self.visit(node.right).value)

    def visitMultiplyNode(self, node: MultiplyNode) -> Number:
        return Number(self.visit(node.left).value * self.visit(node.right).value)

    def visitDivideNode(self, node: DivideNode) -> Number:
        left = self.visit(node.left).value
        right = self.visit(node.right).value
        if right == 0.0:
            raise ZeroDivisionError()
        return Number(self.visit(node.left).value / self.visit(node.right).value)

    def visitMinusNode(self, node: MinusNode) -> Number:
        return Number(-self.visit(node.right).value)

    def visitPlusNode(self, node: PlusNode) -> Number:
        return Number(self.visit(node.right).value)
